### PR TITLE
mpi_reduce example added

### PR DESCRIPTION
**Pull request by gihanlaks**
mpi_reduce example added

### CODE REVIEW

**Comment by gihanlaks** [comment]
@vibhatha @dulvinw for your review

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -50,3 +50,4 @@ endfunction()
 # Example usage
 create_target(hello_world "src/hello_world.cpp")
 create_target(send_recv "src/send_recv.cpp")
+create_target(mpi_reduce "src/mpi_reduce.cpp")
```
Comment (vibhatha) [inline cpp/CMakeLists.txt]: Let's just call it reduce to keep the naming consistent as we are just using MPI.

File: cpp/src/mpi_reduce.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        std::cout << "Global sum = " << global_sum << std::endl;
+        std::cout << "Minimum rank sum = " << global_min << std::endl;
+        std::cout << "Maximum rank sum = " << global_max << std::endl;
+    }
+
+    std::cout << "Local sum = " << local_sum << " for the rank = " << world_rank << std::endl;
+
+    MPI_Finalize();
+    return 0;
+}
```
Comment (vibhatha) [inline cpp/src/mpi_reduce.cpp]: Let's add a new line at the end of the file.

File: cpp/src/mpi_reduce.cpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    int global_max;
+    MPI_Reduce(&local_sum, &global_max, 1, MPI_INT, MPI_MAX, 0, MPI_COMM_WORLD);
+
+    if (world_rank == 0) {
+        std::cout << "Global sum = " << global_sum << std::endl;
+        std::cout << "Minimum rank sum = " << global_min << std::endl;
+        std::cout << "Maximum rank sum = " << global_max << std::endl;
+    }
+
+    std::cout << "Local sum = " << local_sum << " for the rank = " << world_rank << std::endl;
```
Comment (vibhatha) [inline cpp/src/mpi_reduce.cpp]: To follow the logic let's print local min and max too?

**Review by vibhatha — COMMENTED** [review]
Thanks @gihanlaks for making this PR. I added a few comments.

File: cpp/src/mpi_reduce.cpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
+#include <mpi.h>
+#include <iostream>
+
+int main(int argc, char** argv) {
+    MPI_Init(NULL, NULL);
+
+    int world_rank;
+    MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
+
+    int world_size;
```
Comment (dulvinw) [inline cpp/src/mpi_reduce.cpp]: Looks like unused?

File: cpp/src/mpi_reduce.cpp (hunk 3 above)
Comment (gihanlaks) [reply cpp/src/mpi_reduce.cpp]: Here, "Maximum rank sum" means max local sum from all ranks. So LOCAL SUM == LOCAL MAX SUM is true.

File: cpp/src/mpi_reduce.cpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    int local_sum = 0;
+    for (int num : numbers) {
+        local_sum += num;
+    }
+
+    int global_sum;
+    MPI_Reduce(&local_sum, &global_sum, 1, MPI_INT, MPI_SUM, 0, MPI_COMM_WORLD);
+
+    int global_min;
+    MPI_Reduce(&local_sum, &global_min, 1, MPI_INT, MPI_MIN, 0, MPI_COMM_WORLD);
```
Comment (vibhatha) [inline cpp/src/mpi_reduce.cpp]: Oh wait. shouldn't there be a `local_min` variable as well as `local_max` variable?

File: cpp/src/mpi_reduce.cpp (hunk 3 above)
Comment (vibhatha) [reply cpp/src/mpi_reduce.cpp]: I assumed what we are doing here is finding the global min, max and sum.

Comment (vibhatha) [reply cpp/src/mpi_reduce.cpp]: @dulvinw could you please check, I am not sure if I am missing something here.

File: cpp/CMakeLists.txt (hunk 1 above)
Comment (gihanlaks) [reply cpp/CMakeLists.txt]: file name & its references renamed to reduce.cpp

File: cpp/src/mpi_reduce.cpp (hunk 2 above)
Comment (gihanlaks) [reply cpp/src/mpi_reduce.cpp]: New Line added to the end of the file

File: cpp/src/mpi_reduce.cpp (hunk 3 above)
Comment (gihanlaks) [reply cpp/src/mpi_reduce.cpp]: I see your confusion. Your expected output is not what I wanted do here. But your expected output is better. 
So Changed the code to achieve that.

File: cpp/src/mpi_reduce.cpp (hunk 4 above)
Comment (gihanlaks) [reply cpp/src/mpi_reduce.cpp]: world_size was unused. Removed that part.

File: cpp/src/mpi_reduce.cpp (hunk 5 above)
Comment (gihanlaks) [reply cpp/src/mpi_reduce.cpp]: Now there will be those two variables to achieve your expectation.

**Review by dulvinw — APPROVED** [review]

**Review by vibhatha — APPROVED** [review]
@gihanlaks nice work!

**Pull request merged**